include/revng/ADT/Concepts.h

```diff
@@ -4,44 +4,178 @@
+// clang-format on
+
 namespace ranges {
 
 template<class T>
-concept range = requires(T &R) {
+concept range = requires(T &&R) {
   std::begin(R);
   std::end(R);
 };
 
 template<class T>
-concept sized_range = ranges::range<T> && requires(T &R) {
+concept sized_range = ranges::range<T> && requires(T &&R) {
   std::size(R);
 };
 
 } // namespace ranges
 
+//
+// Concepts to simplify working with tuples.
+//
+
+template<class T>
+concept TupleSizeCompatible = requires {
+  std::tuple_size<T>::value;
+  { std::tuple_size_v<T> } -> convertible_to<size_t>;
+};
+
+static_assert(TupleSizeCompatible<std::tuple<>>);
+static_assert(!TupleSizeCompatible<std::vector<int>>);
+static_assert(!TupleSizeCompatible<int>);
+
+namespace revng::detail {
+
+template<class T, std::size_t N>
+concept TupleElementCompatibleHelper = requires(T Value) {
+  typename std::tuple_element_t<N, std::remove_const_t<T>>;
+  { get<N>(Value) } -> convertible_to<std::tuple_element_t<N, T> &>;
+};
+
+template<typename T, size_t... N>
+constexpr auto checkTupleElementTypes(std::index_sequence<N...>) {
+  return (TupleElementCompatibleHelper<T, N> && ...);
+}
+
+template<TupleSizeCompatible T>
+constexpr auto checkAllTupleElementTypes() {
+  constexpr std::size_t Size = std::tuple_size_v<T>;
+  return checkTupleElementTypes<T>(std::make_index_sequence<Size>());
+}
+
+} // namespace revng::detail
+
+// clang-format off
+template<class T>
+concept TupleLike = (TupleSizeCompatible<T>
+                     and revng::detail::checkAllTupleElementTypes<T>());
+// clang-format on
+
+static_assert(TupleLike<std::tuple<>>);
+static_assert(TupleLike<std::tuple<int, int, long>>);
+static_assert(TupleLike<std::pair<int, int>>);
+static_assert(TupleLike<std::array<int, 0>>);
+static_assert(not TupleLike<std::vector<int>>);
+
+//
+// Concepts to simplify working with specializations of templates.
+//
+
+/// A concept that helps determine whether a given object is a specialization
+/// (or is inheriting a specialization) of a given template.
+///
+/// This lets us cut down on the number of `IsX` (e.g. `IsGenericGraph`,
+/// `IsRank`) concepts needed since we can now just say
+/// `SpecializationOf<GenericGraph>` on the interface boundaries and get
+/// the expected check.
+///
+/// TODO: this requires clang-13+, uncomment it after the update.
+// template<typename Type, template<typename...> class Ref>
+// concept SpecializationOf = requires(Type &&Value) {
+//   []<typename... Ts>(Ref<Ts...> &){}(Value);
+// };
+//
+// template<typename Type, template<typename...> class Ref>
+// concept NotSpecializationOf = !SpecializationOf<Type, Ref>;
+//
+// static_assert(SpecializationOf<std::pair<int, long>, std::pair>);
+// static_assert(SpecializationOf<const std::pair<int, long>, std::pair>);
+// static_assert(SpecializationOf<std::string, std::basic_string>);
+// static_assert(SpecializationOf<std::ostream, std::basic_ios>);
+// static_assert(not SpecializationOf<std::string, std::basic_string_view>);
```

Comment:
`SpecializationOf` or `StrictSpecializationOf`?

Reply:
`SpecializationOf` requires `clang-13 :(` ([this paper](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0315r4.pdf)), so for now I used `StrictSpecializationOf` everywhere (it seems to do the trick) and left the comments in a couple places to make it less strict after the update.